pdbfixer/hbonds.py

```diff
@@ -0,0 +1,293 @@
+import openmm as mm
+import openmm.app as app
+import openmm.unit as unit
+from openmm.app.element import hydrogen, oxygen, nitrogen, fluorine
+import numpy as np
+import itertools
+from collections import defaultdict
+from collections.abc import Iterator
+from typing import List
+
+class HydrogenBondOptimizer(object):
+    """This class rotates sidechains to optimize the hydrogen bond network."""
+    
+    def __init__(self, fixer):
+        topology, positions = self.protonateHistidines(fixer)
+
+        # Record the parent atom that every hydrogen is bonded to.
+
+        parent = {}
+        for atom1, atom2 in topology.bonds():
+            if atom1.element == hydrogen and atom2.element != hydrogen:
+                parent[atom1] = atom2
+            elif atom2.element == hydrogen and atom1.element != hydrogen:
+                parent[atom2] = atom1
+
+        # Record the donors and acceptors in each residue.
+
+        self.residueDonors = {}
+        self.residueAcceptors = {}
+        for residue in topology.residues():
+            self.residueDonors[residue] = []
+            self.residueAcceptors[residue] = []
+            for atom in residue.atoms():
+                if atom.element == hydrogen and atom in parent and parent[atom].element in (oxygen, nitrogen, fluorine):
```

Comment:
We might need to include sulfur to capture Cysteine sidechains. Do we need fluorine in this list? H-F bonds are unlikely.

Reply:
Fluorine is one of the most common hydrogen bond acceptors.  It's very common in drug molecules.

Reply:
Sorry, the fluorine comment was meant for the donor list. Definitely keep it for the acceptors 😄

Reply:
According to [Wikipedia](https://en.wikipedia.org/wiki/Hydrogen_bond), F is also a very common donor.

Reply:
Yes, but to be a donor, F needs to be bound to hydrogen. F can only make one bond, so it would be an isolated H-F molecule.